source/mesh_refinement/isosurfaces.cc

```diff
@@ -0,0 +1,391 @@
+
+              for (unsigned int i_quad=0; i_quad<quadrature.size(); ++i_quad)
+                {
+                  for (auto &isosurface : isosurfaces)
+                    {
+                      // setup the vector to check
+                      std::vector<double> values(isosurface.min_values.size());
+                      for (unsigned int index = 0; index < isosurface.properties.size(); ++index)
+                        {
+                          if (isosurface.properties[index].name == Internal::PropertyName::Temperature)
+                            {
+                              values[index] = in.temperature[i_quad];
+                            }
+                          else if (isosurface.properties[index].name == Internal::PropertyName::Composition)
+                            {
+                              values[index] = std::min(std::max(in.composition[i_quad][isosurface.properties[index].index], 0.0), 1.0);
+                            }
+                          else if (isosurface.properties[index].name == Internal::PropertyName::Background)
+                            {
+                              double sum_composition = 0.0;  // Compute background material fraction
+                              for (unsigned i=0 ; i < in.composition[i_quad].size() ; ++i)
+                                {
+                                  sum_composition += std::min(std::max(in.composition[i_quad][i], 0.0), 1.0);
+                                }
+                              if (sum_composition >= 1.0)
+                                values[index] = 0.0;
+                              else
+                                values[index] = 1.0 - sum_composition;
+                            }
+                        }
+
+                      if ( isosurface.values_are_in_range(values))
+                        {
+                          // If the current refinement level is smaller or equal to the minimum
+                          // refinement level, any coarsening flags should be cleared.
+                          if (cell->level() <= isosurface.min_refinement)
+                            {
+                              clear_coarsen = true;
+                            }
+
+                          // If the current refinement level is smaller then the minimum
+                          // refinement level, a refinment flag should be placed.
+                          if (cell->level() <  isosurface.min_refinement)
+                            {
+                              refine = true;
+                              break;
+                            }
+
+                          // If the current refinement level is larger or equal to the maximum refinement
+                          // level, any refinement flag should be cleared.
+                          if (cell->level() >= isosurface.max_refinement)
+                            {
+                              clear_refine = true;
```

Comment:
What happens if two "isosurfaces" overlap. I assume that the "last" one wins? Is that desired?

Reply:
Yes. I don't think there is a  most desired way of doing it, but it is a clear way. I have added a sentence about that in the declare_entry of isosurfaces.